ffi/src/lib.rs

```diff
@@ -388,14 +402,98 @@ fn create_client_internal(
             let redis_error = err.into();
             errors::error_message(&redis_error)
         })?;
+
+    // Wrap the runtime in an Arc so it can be shared
+    let runtime = Arc::new(runtime);
+
     let client = runtime
-        .block_on(GlideClient::new(ConnectionRequest::from(request), None))
+        .block_on(GlideClient::new(
+            ConnectionRequest::from(request),
+            Some(push_tx),
+        ))
         .map_err(|err| err.to_string())?;
+
+    // Create the client adapter that will be returned and used as conn_ptr
     let core = Arc::new(CommandExecutionCore {
-        client,
-        client_type,
+        client: client.clone(),
+        client_type: client_type.clone(),
+    });
+    let client_adapter = Arc::new(ClientAdapter {
+        runtime: runtime.clone(),
+        core,
     });
-    Ok(ClientAdapter { runtime, core })
+    // Clone client_adapter before moving it into the async block
+    let client_adapter_clone = Arc::into_raw(client_adapter.clone()).addr();
+
+    // Check if the pubsub_callback function pointer is not null
+    // We can't directly use is_null() on function pointers, so we need to compare with std::ptr::null()
+    let null_fn: PubSubCallback = unsafe { std::mem::transmute(std::ptr::null::<()>()) };
+    let is_callback_valid = !std::ptr::eq(
+        pubsub_callback as *const () as *const u8,
+        null_fn as *const () as *const u8,
+    );
+
+    // If pubsub_callback is provided (not null), spawn a task to handle push notifications
+    if is_callback_valid {
+        runtime.spawn(async move {
+            loop {
+                let result = push_rx.recv().await;
+                match result {
+                    None => {
+                        return;
+                    }
+                    Some(push_msg) => {
+                        // Convert the push_msg.data to a CommandResponse
+                        let data_response = match valkey_value_to_command_response(Value::Array(
```

Comment:
Message is always an array of 2 or 3 elements. You can change the callback to
```rs
pub type PubSubCallback =
    unsafe extern "C" fn(client_ptr: usize, kind: u32, message: *const u8, message_len: u32, channel: *const u8, channel_len: u32, pattern: *const u8, pattern_len: u32) -> ();
```
And get rid of `valkey_value_to_command_response` and `defer C.free_command_response(cResponse)`. This will give a performance boost, since code won't recursively traverse the structure. Keep in mind message could be a non-utf8 string (channel probably too).

Alternative option is to define a struct with 3/6 fields and marshal it. If you copy it in sync manner in the callback, you don't need to worry about memory freeing too.

Reply:
Have applied the changes you submitted.